docs/madr/decisions/046-meshservice-policy-matching.md

```diff
@@ -0,0 +1,228 @@
+for matching real `MeshService` resources to the existing `MeshService` kind.
+Matches on real `MeshService` objects have priority over `kuma.io/service`
+matches.
+
+A majority of the use cases for policy will involve selecting a specific *port* of the
+`MeshService`, `MeshServices` from a given Kuma zone and namespace.
+
+Note, we have already in targetRef:
+
+```yaml
+name: ...
+namespace: ...
+```
+
+Now we have to consider how we attach to a port and select from a zone.
+
+* Add top level field `zone` and `port` and `labels` for selecting groups of
+  `MeshService`
+* Add top level `port` field and use `labels: {"kuma.io/zone": "..."}`
+* Use generic Gateway API targetRef field `sectionName` to specify a port in
+  `targetRef` but `port` in `backendRef` and add `labels` for selecting
+  groups of `MeshService`.
+
+## Decision Outcome
+
+We choose:
+
+* use generic Gateway API targetRef field `sectionName` to specify `port` and
+  add `labels` but use `port` in `backendRefs`.
+
+### `targetRef`
+
+The policy is applied
+on both the Envoy outbound listeners and clusters
+that correspond to the `MeshService`.
+
+References for `kind: MeshService` have the following structure:
+
+```yaml
+---
+kind: MeshService
+spec:
+  ports:
+  - port: 80
+    name: http
+---
+targetRef:
+  kind: MeshService
+  namespace: backend
+  name: backend # only name or labels
+  labels: # only name or labels
+    kuma.io/zone: east
+  sectionName: http
+```
+
+The `name` of a `MeshService` ref always refers to the name of one, real Kubernetes
+object in the zone of the policy.
+
+Exactly one of either `targetRef.name` or `targetRef.labels` is **required**.
+
+#### `sectionName`
+
+If `sectionName` is set in `targetRef.kind` with `kind: MeshService`,
+it refers to an entry in `ports` by name and only traffic to that port is affected.
+
+In `backendRefs`, `port` is required. There is no `sectionName`, in accordance
+with Gateway API HTTPRoute.
+
+#### `name`
+
+If the policy ref sets `name`, it refers to one `MeshService` in the *local
+zone*.
+
+On a k8s zone:
+* without `namespace`, the policy refers to `MeshServices` in the same
+  namespace as the policy.
+* with `namespace`, only that namespace is searched for matching `MeshServices`
+
+On a universal zone, `namespace` is not valid.
+
+Note that this means it's possible to refer to synced `MeshServices` via their
+transformed `<service>-<hash-suffix>`.
+
+#### `labels`
+
+If `name` is not set, `labels` must be set.
+Via `labels`, groups of `MeshServices` can be matched, for example by:
+
+* `kuma.io/display-name`
+* `kuma.io/zone`
+
+If `kuma.io/zone` is set, a missing `namespace` field means we
+refer to _all_ `MeshServices` with this name.
+If no `kuma.io/zone` is set, the `namespace` is assumed to be the namespace of
+the policy making the reference.
+
+This means we don't infer any similarity between the local namespace
+and the other zone's namespace.
```

Comment:
This is not as nice as I thought. So we don't want to assume namespace similarity, that's good. But how do you refer to a MeshService from another zone then?

```
namespace: other-zone-ns
labels:
  kuma.io/zone: other-zone
```

isn't ideal because the real object is not found at this namespace, rather at `kuma-system`.

so do we have

```
namespace: kuma-system
labels:
  kuma.io/zone: other-zone
  k8s.kuma.io/zone: other-zone-ns
```

? This is the cleanest but it's very verbose.

Reply:
Shouldn't it be `k8s.kuma.io/namespace: other-zone-ns` in the second example?

Reply:
So this means that you need to create policy in `kuma-system` namespace in order to target `MeshService` in different zone? While discussing policies on namespace MADR we talked that in best case scenario SO won't need to apply anything on `kuma-system` namespace

Reply:
Well, TBH it's not drastically more verbose than the first examples. I'd still with the more verbose version in favor of preserving the rule that we reference real objects

Reply:
I agree with Ilya here ☝️

Reply:
OK spoke with @jakubdyszkiewicz and we decided to just disallow `namespace` when using `labels`, and all namespaces will be matched. This makes it more verbose for the local case of a group of MeshServices but :shrug:

Reply:
> So this means that you need to create policy in kuma-system namespace in order to target MeshService in different zone?

I don't think so, it's just about `namespace` in the target ref itself. Updated the YAML to make it clearer.